src/Common/FailPoint.cpp

```diff
@@ -0,0 +1,154 @@
+#include <Common/Exception.h>
+#include <Common/FailPoint.h>
+
+#include <boost/core/noncopyable.hpp>
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+#include <optional>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int LOGICAL_ERROR;
+};
+
+#if FIU_ENABLE
+static struct InitFiu
+{
+    InitFiu()
+    {
+        fiu_init(0);
+    }
+} init_fiu;
+#endif
+
+#define APPLY_FOR_FAILPOINTS(ONCE, REGULAR, PAUSEABLE_ONCE, PAUSEABLE) \
```

Comment:
A comment explaining how to add new failure points and what are the differences between failpoint types is required here.